File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/strategies/rule_based.py
from __future__ import annotations

from elizaos_plugin_auto_trader.types import MarketData, TradeSignal


def _moving_average(prices: list[float], window: int) -> float | None:
    if len(prices) < window or window == 0:
        return None
    s = prices[-window:]
    return sum(s) / len(s)


class RuleBasedStrategy:
    """Simple moving-average crossover strategy (simulated).

    Uses a short and long window over the price history to detect crossovers.
    Short MA above long MA → buy signal.
    Short MA below long MA → sell signal.
    """

    def __init__(self, short_window: int = 5, long_window: int = 20) -> None:
        self.short_window = max(1, short_window)
        self.long_window = max(2, long_window)

    @property
    def name(self) -> str:
        return "RuleBased"

    async def analyze(self, market_data: MarketData) -> TradeSignal | None:
        prices = market_data.prices
        if len(prices) < self.long_window:
            return None

        short_ma = _moving_average(prices, self.short_window)
        long_ma = _moving_average(prices, self.long_window)

        if short_ma is None or long_ma is None or abs(long_ma) < 1e-12:
            return None

        diff_pct = (short_ma - long_ma) / long_ma * 100.0

        if diff_pct > 1.0:
            return TradeSignal(
                token=market_data.token,
                direction="BUY",
                strength=min(diff_pct / 10.0, 1.0),
                reason=(
                    f"SMA crossover: short({self.short_window})={short_ma:.2f} "
                    f"> long({self.long_window})={long_ma:.2f} ({diff_pct:+.2f}%)"
                ),
            )
        elif diff_pct < -1.0:
            return TradeSignal(
                token=market_data.token,
                direction="SELL",
                strength=min(abs(diff_pct) / 10.0, 1.0),
                reason=(
                    f"SMA crossover: short({self.short_window})={short_ma:.2f} "
                    f"< long({self.long_window})={long_ma:.2f} ({diff_pct:+.2f}%)"
                ),
            )
        return None

File: plugins/plugin-auto-trader/rust/src/strategies/rule_based.rs
use async_trait::async_trait;

use super::Strategy;
use crate::types::{MarketData, TradeDirection, TradeSignal};

/// Simple moving-average crossover strategy (simulated).
///
/// Uses a short and long window over the price history to detect crossovers.
/// When the short MA crosses **above** the long MA → buy signal.
/// When the short MA crosses **below** the long MA → sell signal.
pub struct RuleBasedStrategy {
    pub short_window: usize,
    pub long_window: usize,
}

impl Default for RuleBasedStrategy {
    fn default() -> Self {
        Self {
            short_window: 5,
            long_window: 20,
        }
    }
}

impl RuleBasedStrategy {
    pub fn new(short_window: usize, long_window: usize) -> Self {
        Self {
            short_window: short_window.max(1),
            long_window: long_window.max(2),
        }
    }
}

fn moving_average(prices: &[f64], window: usize) -> Option<f64> {
    if prices.len() < window || window == 0 {
        return None;
    }
    let slice = &prices[prices.len() - window..];
    Some(slice.iter().sum::<f64>() / window as f64)
}

#[async_trait]
impl Strategy for RuleBasedStrategy {
    fn name(&self) -> &str {
        "RuleBased"
    }

    async fn analyze(&self, market_data: &MarketData) -> Option<TradeSignal> {
        let prices = &market_data.prices;
        if prices.len() < self.long_window {
            return None;
        }

        let short_ma = moving_average(prices, self.short_window)?;
        let long_ma = moving_average(prices, self.long_window)?;

        // Avoid divide-by-zero
        if long_ma.abs() < f64::EPSILON {
            return None;
        }

        let diff_pct = (short_ma - long_ma) / long_ma * 100.0;

        if diff_pct > 1.0 {
            // Short MA above long MA → bullish crossover
            Some(TradeSignal {
                token: market_data.token.clone(),
                direction: TradeDirection::Buy,
                strength: (diff_pct / 10.0).min(1.0),
                reason: format!(
                    "SMA crossover: short({})={:.2} > long({})={:.2} ({:+.2}%)",
                    self.short_window, short_ma, self.long_window, long_ma, diff_pct
                ),
            })
        } else if diff_pct < -1.0 {
            // Short MA below long MA → bearish crossover
            Some(TradeSignal {
                token: market_data.token.clone(),
                direction: TradeDirection::Sell,
                strength: (diff_pct.abs() / 10.0).min(1.0),
                reason: format!(
                    "SMA crossover: short({})={:.2} < long({})={:.2} ({:+.2}%)",
                    self.short_window, short_ma, self.long_window, long_ma, diff_pct
                ),
            })
        } else {
            None
        }
    }
}
